File: LeetCode/015ThreeSum/Solution.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        if (nums == null || nums.length < 3) {
            return result;
        }

        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 2; i++) {
            // 裁枝
            if (nums[i] > 0) {
                break;
            }
            // 去重
            if (i > 0 && nums[i] == nums[i-1]) {
                continue;
            }
            int left = i + 1;
            int right = nums.length - 1;
            while (left < right) {
                int tmp = nums[i] + nums[left] + nums[right];
                if (tmp < 0 || (left > i+1 && nums[left] == nums[left-1])) {
                    left++;
                } else if (tmp > 0 || (right < nums.length-1 && nums[right] == nums[right+1])) {
                    right--;
                } else {
                    List<Integer> list = new ArrayList<Integer>();
                    list.add(nums[i]);
                    list.add(nums[left++]);
                    list.add(nums[right--]);
                    result.add(list);
                }
            }
        }

        return result;
    }
}

File: LeetCode/015ThreeSum/Solution.py
class Solution:
    def threeSum1(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3:
            return []
        res = set()
        for i, a in enumerate(nums[:-2]):
            target = set()
            for b in nums[i + 1:]:
                if b not in target:
                    target.add(-a - b)
                else:
                    res.add(tuple(sorted([a, -a - b, b])))
        return map(list, res)

    def threeSum2(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3:
            return []
        nums.sort()
        res = set()
        for i, v in enumerate(nums[:-2]):
            if v > 0:
                break
            if i >= 1 and v == nums[i - 1]:
                continue
            target = set()
            for x in nums[i + 1:]:
                if x not in target:
                    target.add(-v - x)
                else:
                    res.add((v, -v - x, x))
        return map(list, res)

    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3:
            return []
        nums.sort()
        res = []
        for i, v in enumerate(nums[:-2]):
            if v > 0:
                break
            if i > 0 and v == nums[i - 1]:
                continue
            l, r = i + 1, len(nums) - 1
            while l < r:
                s = v + nums[l] + nums[r]
                if s < 0:
                    l += 1
                elif s > 0:
                    r -= 1
                else:
                    res.append([v, nums[l], nums[r]])
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    while l < r and nums[r] == nums[r - 1]:
                        r -= 1
                    l += 1
                    r -= 1
        return res